// nv-array-basic/index.js
function clear(arr) {
    arr.splice(0,Infinity)
    return(arr)
}

function head(arr) {
    return(arr[0])
}

function tail(arr) {
    return(arr.slice(1))
}

function mid(arr) {
    return(arr.slice(1,arr.length-1))
}

function init(arr) {
    return(arr.slice(0,arr.length-1))
}


function lst(arr) {
    return(arr[arr.length-1])
}

function validate_index(arr,i) {
    let cond = (i>=0) && (i<arr.length)
    if(cond) {
        return(i)
    } else {
        return(undefined)
    }
}

function validate_value(arr,v) {
    if(arr.indexOf(v)>=0) {
        return(v)
    } else {
        return(undefined)
    }
}

function validate_vi(arr,d) {
    let i = d.i
    let v = d.v
    if(validate_index(arr,i)!==undefined) {
        if(arr[i] === v) {
            return(d)
        } else {
            return(undefined)
        }
    } else {
        return(undefined)
    }
}


function some(arr,...whiches) {
    return(arr.filter((r,i)=>whiches.includes(i)))
}

function some_not(arr,...whiches) {
    return(arr.filter((r,i)=>!whiches.includes(i)))
}

function interval(arr,interval) {
    return(arr.filter((r,i)=>i%interval===0))
}

function odd(arr) {
    return(arr.filter((r,i)=>i%2===1))
}
function even(arr) {
    return(arr.filter((r,i)=>i%2===0))
}


module.exports = {
    clear,
    head,
    tail,
    mid,
    init,
    lst,
    validate_index,
    validate_value,
    validate_vi,
    some,
    some_not,
    interval,
    odd,
    even
}
